### PR TITLE
Completed stack language, testing included

### PR DESCRIPTION
**Pull request by osu-mp**
Operations and syntax are in the shared file, testing/demonstration is in another file that we won't turn in

### CODE REVIEW

**Review by psulliv — APPROVED** [review]
```
>./3_StackLang 
./3_StackLang 

Test - expect failure on second add
Error: ADD requires at least 2 elements on stack

Test - Expect [6, 12] via  (3; 3,3; 6; 6,6; 6,6,6; 12,6; 6, 12)
[6,12]

Test - Expect [] (no operations)
[]

Test - Expect [10,1,5] via (1; 5,1; 1,5; 1,1,5; 1,1,1,5; 2,1,5; 8,2,1,5; 10,1,5)
[10,1,5]

Test - Expect failure when duplicating empty stack
Error: DUP requires at least 1 element on stack

DONE
```

Ship it!

File: hw3/hw3_group2.hs [hunk 1]
```diff
@@ -1,3 +1,39 @@
...
+
+-- result of operation can be either a stack or a failure string
+data Result a = Success a | Fail String
+semOp :: Op -> Stack -> Result Stack
+
+-- load given int to top of stack
+semOp (LD x) stack = Success (x : stack)
+
+-- add the first 2 elements on the stack or error if fewer than 2 elements present
+semOp ADD [] = Fail ("Error: empty stack, cannot add")
```
Comment (caitlynplewis) [inline hw3/hw3_group2.hs]: You don't strictly need the empty list case for this or SWAP - the _ case will catch anything that doesn't match the two-element case, including empty.

File: hw3/hw3_group2.hs [hunk 2]
```diff
@@ -1,3 +1,39 @@
...
+semOp SWAP [] = Fail ("Error: swap requires 2 elements on stack")
+semOp SWAP (x:y:stack) = Success (y : x: stack)
+semOp SWAP _ = Fail ("Error: SWAP requires at least 2 elements on stack")
+
+-- duplicate top element and push onto stack
+-- if stack is empty, produce error
+semOp DUP [] = Fail ("Error: DUP requires at least 1 element on stack")
+semOp DUP (x:xs) = Success (x:x:xs)
+
+semS :: Stack -> [Op] -> Result Stack
```
Comment (caitlynplewis) [inline hw3/hw3_group2.hs:32]: It's probably technically more correct to have [Op] -> Stack ->Result Stack, to preserve the [Op] -> D pattern he asks for, though practically it makes no difference.

**Review by caitlynplewis — APPROVED** [review]
👍

File: hw3/hw3_group2.hs (hunk 1 above)
Comment (osu-mp) [reply hw3/hw3_group2.hs]: Good catch, removed both.

File: hw3/hw3_group2.hs (hunk 2 above)
Comment (osu-mp) [reply hw3/hw3_group2.hs:32]: Making that addition would require changing a few other things to make it fit, so I'll leave it as is for now unless either of you feel strongly about the change.

**Pull request merged**